### PR TITLE
feat(linter/release): automate oxlint rule version updates

### PR DESCRIPTION
**Pull request by penkzhou**
## Summary

This PR covers the release automation follow-up from #19890.

It replaces the earlier Rust-based prototype with a small Node script, `.github/scripts/update-rule-versions.js`, that scans `declare_oxc_lint!` declarations under `crates/oxc_linter/src/rules/**` and:

- rewrites stable `version = "next"` entries to the actual oxlint release version during release prep
- leaves `nursery` rules on `version = "next"`
- reuses the same script during release CI and fails if rerunning it would still rewrite tracked rule files, so no stable `version = "next"` remains at release time
- prints a simple change report
- supports `--dry-run` for local verification

The implementation also now hardens a few failure modes that came up during review:

- comment-only lines inside `declare_oxc_lint!` blocks are ignored when parsing metadata
- trailing `// ...` and single-line `/* ... */` comments on metadata lines are handled correctly
- multi-line block comments on metadata lines and macro terminators are handled correctly
- doc examples inside the macro block that mention `version = "next"` are not misclassified as invalid matches
- symlinked entries under the lint rules tree are rejected instead of being silently skipped
- the script still fails if `version = "next"` is found outside a `declare_oxc_lint!` block

Workflow changes:

- `.github/workflows/prepare_release_apps.yml` sets up Node and runs the rewrite script after `cargo release-oxc update`
- `.github/workflows/release_apps.yml` sets up Node, reruns the script with the released oxlint version, and then checks `git diff --exit-code -- crates/oxc_linter/src/rules`

## Testing

```bash
node --test .github/scripts/update-rule-versions.test.js
node .github/scripts/update-rule-versions.js --root . --release-version 1.61.0 --dry-run
cargo check -p oxc_linter --locked
```

## Notes

This PR intentionally does not include the docs/website wiring from #19890.
That should stay in a follow-up PR since it touches a different data flow and also involves `oxc-project/website#937`.

## AI Usage Disclosure

I used AI assistance to help inspect the release workflows, compare the implementation direction against the maintainer feedback, and draft the initial automation changes. I manually reviewed the final implementation, added the regression tests for the workflow guard and parser edge cases above, ran the checks listed here, and remain responsible for the final code and submission.

### CODE REVIEW

**Comment by penkzhou** [comment]
I put the release-version logic in a small `tasks/` crate rather than directly in workflow YAML so it stays testable, debuggable locally, and easier to evolve as the rule metadata checks grow.

File: tasks/release_rule_versions/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,665 @@
...
+        }
+
+        file_rewrites.push(FileRewrite::new(path, source, rewrites));
+    }
+
+    for file_rewrite in file_rewrites {
+        ensure_path_is_within_rules_root(&file_rewrite.path, &rules_root)?;
+        fs::write(&file_rewrite.path, file_rewrite.rewritten_source)?;
+        rewritten_rules.extend(file_rewrite.rewritten_rules);
+    }
```
[the remark this answers is not in the document]
Comment (penkzhou) [reply tasks/release_rule_versions/src/lib.rs]: Addressed in e71aa0fd1b.

Changes:
- prepare all rewritten files as same-directory temporary files first
- atomically replace the original files via `NamedTempFile::persist(...)` instead of writing in place
- add a regression test that fails on the old implementation when a later write-stage failure would otherwise leave earlier files already rewritten
- switch the test fixture helper to `TempDir` to avoid parallel-test directory collisions

Verification:
- `cargo test -p oxc_release_rule_versions`
- `cargo run -p oxc_release_rule_versions -- check --root .`
- `cargo check -p oxc_release_rule_versions --all-targets --locked`
- `cargo fmt --all`

This keeps the original tree untouched until all rewritten files have been prepared successfully and avoids truncating target files during the rewrite step.

File: tasks/release_rule_versions/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,753 @@
...
+        file_rewrites.push(FileRewrite::new(path, source, rewrites));
+    }
+
+    for file_rewrite in file_rewrites {
+        ensure_path_is_within_rules_root(&file_rewrite.path, &rules_root)?;
+        prepared_file_rewrites.push(PreparedFileRewrite::new(file_rewrite)?);
+    }
+
+    for prepared_file_rewrite in prepared_file_rewrites {
+        rewritten_rules.extend(prepared_file_rewrite.persist()?);
```
[the remark this answers is not in the document]
Comment (penkzhou) [reply tasks/release_rule_versions/src/lib.rs]: Implemented this with rollback on later persist failures.

The rewrite flow now keeps a same-directory backup temp file for each touched rule file, persists rewritten files in a second phase, and rolls back already-persisted files in reverse order if a later persist fails. I also made rollback best-effort so it still attempts to restore every previously persisted file even if one restore fails.

Added a regression test for this exact case: `rewrite_rolls_back_if_a_later_persist_fails`.

Re-verified with:
- `cargo test -p oxc_release_rule_versions`
- `cargo run -p oxc_release_rule_versions -- check --root .`
- `cargo check -p oxc_release_rule_versions --all-targets --locked`
- `cargo fmt --all`

Pushed in `001da6030b`.

**Review by camchenry — CHANGES_REQUESTED** [review]
Thanks for working on this! This looks in the right direction, but I think this is a bit overcomplicated for what we need. At its core, this is a simple string replacement. We turn `version = "next"` into `version = "1.2.3"`, with some caveats (like handling nursery rules). Some general thoughts on how to make this simpler:

- Let's make this a JavaScript script that we run with node. No Rust source code parsing, so it will be simpler and faster. General workflow:
  - Find .rs files that contain the text `version = "next"`. Maybe do a check it's in the `declare_oxc_lint!` block too.
  - For those files, check if the category is set to nursery.
  - If we should update, replace the version text range with the new version text.
- We don't need a separate `check` tool - we only ever call this after we just rewrote the version, so this is redundant. I don't think we need the check in release_apps.yml either - we only need to update it when a release is performed, that should be good enough I think
- Output a simple report of what was changed (list of files modified and what was replaced?)
- Maybe add a `--dry-run` option too or something just in case

**Comment by camchenry** [comment]
I'd recommend taking a look at this equivalent from Biome: https://github.com/biomejs/biome/blob/eaa7d3a7d57f1e08098c94a36aff374e5565d2e4/scripts/update-next-version.sh#L31, but implementing it as a node script instead of shell.

**Comment by penkzhou** [comment]
Alright, I'll take a look. Thank you for your guidance.

File: .github/scripts/update-rule-versions.js [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+    if (metadataEntries.length < 3) {
+      throw new Error(`${relativeFile}: could not parse rule category from declare_oxc_lint! block`);
+    }
+
+    const ruleName = metadataEntries[0].trimmed.replace(/,$/, "");
+    const category = metadataEntries[2].trimmed.replace(/,$/, "");
+
+    if (!VALID_CATEGORIES.has(category)) {
+      throw new Error(`${relativeFile}: unknown rule category \`${category}\``);
```
[the remark this answers is not in the document]
Comment (penkzhou) [reply .github/scripts/update-rule-versions.js]: Fixed.

I now strip trailing `// ...` comments before reading the rule name/category tokens, so declarations like `nursery, // ...` are treated as valid metadata instead of failing category validation.

I also added a regression test for this case: `accepts inline comments on category lines`.

Re-verified with:
- `node --test .github/scripts/update-rule-versions.test.js`
- `node .github/scripts/update-rule-versions.js --root . --release-version 1.61.0 --dry-run`

Pushed in `af2b4a8591`.

File: .github/scripts/update-rule-versions.js [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+function collectRuleFiles(dir) {
+  const files = [];
+
+  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
+    const entryPath = path.join(dir, entry.name);
+    if (entry.isDirectory()) {
+      files.push(...collectRuleFiles(entryPath));
+    } else if (entry.isFile() && entry.name.endsWith(".rs")) {
+      files.push(entryPath);
```
[the remark this answers is not in the document]
Comment (penkzhou) [reply .github/scripts/update-rule-versions.js]: Fixed in 6996b94a21.

I now fail fast on symlinked entries while traversing the rules tree instead of silently skipping them. `collectRuleFiles(...)` rejects `entry.isSymbolicLink()` and reports the offending path relative to the repo root, so release prep cannot leave a symlinked stable rule on `version = "next"` unnoticed.

I also added a regression test for this exact case: `fails if the rules tree contains symlinked rule files`.

Re-verified with:
- `node --test .github/scripts/update-rule-versions.test.js`
- `node .github/scripts/update-rule-versions.js --root . --release-version 1.61.0 --dry-run`

**Comment by penkzhou** [comment]
Thanks again for the simplification suggestion. I updated the PR to follow that direction more closely.

The release-version update is now handled by a small Node script instead of the earlier Rust-based approach. It keeps the flow focused on the actual need here: rewriting stable `version = "next"` entries during release prep, skipping `nursery`, supporting `--dry-run`, and printing a simple report.

I also hardened a couple of edge cases that came up in review:
- comment-only / trailing comments inside `declare_oxc_lint!` blocks are handled correctly
- symlinked rule paths now fail fast instead of being silently skipped

Both of those now have regression tests as well.

When you have time, could you please take another look? cc @camchenry

File: .github/scripts/update-rule-versions.js [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+  for (const [lineIndex, line] of lines.entries()) {
+    const isCommentLineInsideDeclareRuleBlock = declareRuleBlocks.some(
+      ({ startLine, endLine }) =>
+        lineIndex > startLine && lineIndex < endLine && isCommentOnlyLine(line.trim()),
+    );
+    if (
+      NEXT_VERSION_REGEX.test(line) &&
+      !coveredNextVersionLines.has(lineIndex) &&
+      !isCommentLineInsideDeclareRuleBlock
+    ) {
```
[the remark this answers is not in the document]
Comment (penkzhou) [reply .github/scripts/update-rule-versions.js]: Fixed in 85b614e1e8.

The stray-literal validation now scans `stripTrailingComments(line)` instead of the raw line text, so inline comments like `nursery, // keep version = "next"` no longer trigger a false `outside a declare_oxc_lint! block` error.

I also added a regression test for this case: `ignores inline comments in the stray next-version scan`.

Re-verified with `node --test .github/scripts/update-rule-versions.test.js`.

File: .github/scripts/update-rule-versions.js [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+  if (!fs.existsSync(rulesRoot)) {
+    throw new Error(`rules root does not exist: ${rulesRoot}`);
+  }
+
+  const report = { updatedRules: [], skippedNurseryRules: [] };
+
+  for (const filePath of collectRuleFiles(rulesRoot, repoRoot)) {
+    const source = fs.readFileSync(filePath, "utf8");
+    if (!source.includes(NEXT_VERSION_TEXT)) {
+      continue;
```
[the remark this answers is not in the document]
Comment (penkzhou) [reply .github/scripts/update-rule-versions.js]: Fixed in 85b614e1e8.

The early file-level skip now uses `NEXT_VERSION_REGEX.test(source)` instead of the exact substring check, so spacing variants like `version= "next"` and tabbed forms are analyzed and rewritten consistently with the parser/rewrite logic.

I also added a regression test for this case: `rewrites spacing variants of version next`.

Re-verified with:
- `node --test .github/scripts/update-rule-versions.test.js`
- `node .github/scripts/update-rule-versions.js --root . --release-version 1.61.0 --dry-run`

File: .github/scripts/update-rule-versions.js [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+    startLine = endLine;
+  }
+
+  for (const [lineIndex, line] of lines.entries()) {
+    const isCommentLineInsideDeclareRuleBlock = declareRuleBlocks.some(
+      ({ startLine, endLine }) =>
+        lineIndex > startLine && lineIndex < endLine && isCommentOnlyLine(line.trim()),
+    );
+    if (
+      NEXT_VERSION_REGEX.test(line) &&
```
[the remark this answers is not in the document]
Comment (penkzhou) [reply .github/scripts/update-rule-versions.js]: Fixed in 85b614e1e8.

I now strip inline comment text before running the stray `next` scan, so metadata lines with comments that mention `version = "next"` are not misclassified as outside the macro block.

Covered by the new regression test: `ignores inline comments in the stray next-version scan`.

Re-verified with `node --test .github/scripts/update-rule-versions.test.js`.

File: .github/scripts/update-rule-versions.js [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+  const rulesRoot = path.join(repoRoot, DEFAULT_RULES_ROOT);
+  if (!fs.existsSync(rulesRoot)) {
+    throw new Error(`rules root does not exist: ${rulesRoot}`);
+  }
+
+  const report = { updatedRules: [], skippedNurseryRules: [] };
+
+  for (const filePath of collectRuleFiles(rulesRoot, repoRoot)) {
+    const source = fs.readFileSync(filePath, "utf8");
+    if (!source.includes(NEXT_VERSION_TEXT)) {
```
[the remark this answers is not in the document]
Comment (penkzhou) [reply .github/scripts/update-rule-versions.js]: Fixed in 85b614e1e8.

I switched the early file precheck to `NEXT_VERSION_REGEX.test(source)`, so files with valid spacing variants like `version= "next"` are no longer skipped before analysis.

Covered by the new regression test: `rewrites spacing variants of version next`.

Re-verified with `node --test .github/scripts/update-rule-versions.test.js`.

File: .github/workflows/prepare_release_apps.yml [hunk 9]
```diff
@@ -47,7 +49,8 @@
             cat ./target/OXFMT_CHANGELOG
           } > ./target/PR_BODY.md
 
-      - uses: oxc-project/setup-node@4c26e7cb3605b6bdef5450dacd02c434b10fd8ba # v1.2.0
+      - name: Rewrite rule versions
+        run: node .github/scripts/update-rule-versions.js --release-version "${{ steps.run.outputs.OXLINT_VERSION }}"
```
[the remark this answers is not in the document]
Comment (penkzhou) [reply .github/workflows/prepare_release_apps.yml]: Fixed in f8e7889065.

I updated the workflow step to pass `steps.run.outputs.OXLINT_VERSION` through an intermediate environment variable and then reference `"$OXLINT_VERSION"` in the shell command. That follows GitHub's recommended mitigation for script-injection risks in inline workflow scripts.

The step is now:
- `env: OXLINT_VERSION: ${{ steps.run.outputs.OXLINT_VERSION }}`
- `run: node .github/scripts/update-rule-versions.js --release-version "$OXLINT_VERSION"`

File: .github/scripts/update-rule-versions.js [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+    trimmedLine.startsWith("/*") ||
+    trimmedLine.startsWith("*") ||
+    trimmedLine.startsWith("*/")
+  );
+}
+
+function stripTrailingComments(line) {
+  return line
+    .replace(/\s*\/\/.*$/, "")
+    .replace(/\s*\/\*.*\*\/\s*$/, "")
```
[the remark this answers is not in the document]
Comment (penkzhou) [reply .github/scripts/update-rule-versions.js]: Fixed in 89db88fd56.

`stripTrailingComments(...)` now removes trailing `/* ... */` comments before stripping `// ...`, so metadata lines like `nursery, /* see https://example.com */` are normalized correctly instead of being truncated at the URL `//`.

I also added a regression test for this case: `accepts block comments with urls on category lines`.

Re-verified with:
- `node --test .github/scripts/update-rule-versions.test.js`
- `node .github/scripts/update-rule-versions.js --root . --release-version 1.61.0 --dry-run`

File: .github/scripts/update-rule-versions.js [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+      continue;
+    }
+
+    const fileReport = analyzeRuleFile(source, filePath, releaseVersion, repoRoot);
+    report.updatedRules.push(...fileReport.updatedRules);
+    report.skippedNurseryRules.push(...fileReport.skippedNurseryRules);
+
+    if (!dryRun && fileReport.updatedRules.length > 0) {
+      fs.writeFileSync(filePath, fileReport.updatedSource);
+    }
```
[the remark this answers is not in the document]
Comment (penkzhou) [reply .github/scripts/update-rule-versions.js]: Good catch. This was a real issue: the script was writing files during traversal, so a later validation error could leave earlier files already rewritten.

I changed it to collect pending rewrites during the scan and only persist them after the full validation pass completes. I also added a regression test for this exact case: `does not persist earlier rewrites if a later file fails validation`.

Fixed in `e0ef53e143`.

File: .github/scripts/update-rule-versions.js [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+    trimmedLine.startsWith("/*") ||
+    trimmedLine.startsWith("*") ||
+    trimmedLine.startsWith("*/")
+  );
+}
+
+function stripTrailingComments(line) {
+  return line
+    .replace(/\s*\/\*.*\*\/\s*$/, "")
+    .replace(/\s*\/\/.*$/, "")
```
[the remark this answers is not in the document]
Comment (penkzhou) [reply .github/scripts/update-rule-versions.js]: This is a theoretical gap in the current string-based comment stripping, but it depends on a fairly specific input shape: a non-macro Rust string outside `declare_oxc_lint!` that contains both a URL and the exact `version = "next"` literal.

I checked the current `crates/oxc_linter/src/rules` tree and did not find a real instance of that pattern. Given the intended scope of this script is lint-rule metadata rather than general Rust parsing, I’d prefer not to add more parsing complexity here unless we see a concrete in-repo case.

File: .github/scripts/update-rule-versions.js [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+  const updatedRules = [];
+  const skippedNurseryRules = [];
+
+  for (let startLine = 0; startLine < lines.length; startLine++) {
+    if (!lines[startLine].includes(DECLARE_RULE_MACRO)) {
+      continue;
+    }
+
+    let endLine = startLine + 1;
+    while (endLine < lines.length && lines[endLine].trim() !== ");") {
```
[the remark this answers is not in the document]
Comment (penkzhou) [reply .github/scripts/update-rule-versions.js]: Fixed in `f74b7cee04`. The block scan now accepts `); // ...` and `); /* ... */` terminators instead of requiring the line to be exactly `);`. Added a regression test too.

File: .github/scripts/update-rule-versions.js [hunk 14]
```diff
@@ -0,0 +1,316 @@
...
+    report.skippedNurseryRules.push(...fileReport.skippedNurseryRules);
+
+    if (!dryRun && fileReport.updatedRules.length > 0) {
+      pendingWrites.push({ filePath, updatedSource: fileReport.updatedSource });
+    }
+  }
+
+  if (!dryRun) {
+    for (const pendingWrite of pendingWrites) {
+      fs.writeFileSync(pendingWrite.filePath, pendingWrite.updatedSource);
```
[the remark this answers is not in the document]
Comment (penkzhou) [reply .github/scripts/update-rule-versions.js]: I agree rollback on write failures would make this stricter, but I’m not planning to add that here for now. The remaining gap is only on write-time failures like `ENOSPC`/`EPERM`, and adding backup/rollback machinery feels too heavy for this release helper relative to that risk.

**Comment by penkzhou** [comment]
Implemented the requested follow-up in the current PR.

  What changed:
  - `prepare_release_apps.yml` now runs `.github/scripts/update-rule-versions.js` after `cargo release-oxc update`, so stable `version = "next"` entries are rewritten to the actual oxlint release version during
  release prep.
  - `nursery` rules are still left on `version = "next"` and are reported as skipped instead of being rewritten.
  - `release_apps.yml` now reuses the same script during release CI and then runs `git diff --exit-code -- crates/oxc_linter/src/rules`, so the workflow fails if any stable rule would still be rewritten at release
  time.

  I also hardened the script a bit further while keeping the implementation string-based and Node-only:
  - it now handles multi-line block comments on metadata lines and on commented macro terminators
  - it still rejects stray `version = "next"` outside `declare_oxc_lint!`
  - existing edge-case coverage for inline comments, doc examples, and symlinked rule paths remains in place

  Verification:
  - `node --test .github/scripts/update-rule-versions.test.js`
  - `node .github/scripts/update-rule-versions.js --root . --release-version 1.61.0 --dry-run`
  - `cargo check -p oxc_linter --locked`

cc @camchenry @camc314

File: .github/workflows/release_apps.yml [hunk 15]
```diff
@@ -26,6 +26,9 @@ jobs:
       oxlint_version: ${{ steps.oxlint.outputs.version }}
     steps:
       - uses: taiki-e/checkout-action@7d1e50e93dc4fb3bba58f85018fadf77898aee8b # v1.4.2
+
+      - uses: oxc-project/setup-node@4c26e7cb3605b6bdef5450dacd02c434b10fd8ba # v1.2.0
```
Comment (camc314) [inline .github/workflows/release_apps.yml:30]: if we are setting up node, we should rewrite the script in TS - our node version is at 24 iirc, so supports running TS natively

Comment (penkzhou) [reply .github/workflows/release_apps.yml:30]: Great idea, I'll adjust it.

**Review by camchenry — APPROVED** [review]
ok @camc314 I think this is good to go, if you want to take another look too since it's release related. I've tried to really simplify this down to only the things we actually need since we probably don't want any complex workflows in our release pipeline.

to that end, the tests were helpful while developing this, but I don't think we'll need them long-term, so I deleted them. at this point, I don't expect us to need to make lots of changes that justifies running the tests all the time. keeps things simpler too.

**Review by camc314 — APPROVED** [review]
Thanks - looks good to me!
